controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -140,30 +144,108 @@ public String getScopeName() {
         return deleteStream();
     }
 
+    @Override
+    public CompletableFuture<Void> startTruncation(final Map<Integer, Long> streamCut) {
+        return FutureHelpers.allOfWithResults(streamCut.keySet().stream().map(x -> getSegment(x).thenApply(segment ->
+                new SimpleEntry<>(segment.keyStart, segment.keyEnd)))
+                .collect(Collectors.toList()))
+                .thenAccept(list -> {
+                    // verify that stream cut covers the entire range of 0.0 to 1.0 keyspace without overlaps.
+                    List<SimpleEntry<Double, Double>> reduced = TableHelper.reduce(list);
+                    Exceptions.checkArgument(reduced.size() == 1 && reduced.get(0).getKey().equals(0.0) &&
+                                    reduced.get(0).getValue().equals(1.0), "streamCut",
+                            " Invalid input, Stream Cut does not cover full key range.");
+                }).thenCompose(valid -> getTruncationData(true)
+                        .thenCompose(truncationData -> {
+                            Preconditions.checkNotNull(truncationData);
+                            StreamProperty<StreamTruncationRecord> previous = SerializationUtils.deserialize(truncationData.getData());
+                            Exceptions.checkArgument(!previous.isUpdating(), "TruncationRecord", "Truncation record conflict");
+
+                            return computeTruncationRecord(previous.getProperty(), streamCut)
+                                    .thenApply(StreamProperty::update)
+                                    .thenCompose(prop -> setTruncationData(
+                                            new Data<>(SerializationUtils.serialize(prop), truncationData.getVersion())));
+                        }));
+    }
+
+    private CompletableFuture<StreamTruncationRecord> computeTruncationRecord(StreamTruncationRecord truncationRecord,
+                                                                              Map<Integer, Long> streamCut) {
+        log.debug("computing truncation for stream {}/{}", scope, name);
+        return getHistoryTableFromStore()
+                .thenCompose(history -> getSegmentTableFromStore()
+                        .thenCompose(segment -> getIndexTable()
```

Comment:
These can chain rather than nest

Reply:
Nesting is required because we want to use all the collected metadata in the leaf node.. 
with chaining we will need to keep creating tuples and passing on to next element in the chain. 
or have some final atomic reference which we set and make use of in the leaf node where we make use of all the collected data.